### PR TITLE
Add special feature for inter-crate safe dev-related code reuse

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

I'm tired of waiting https://github.com/rust-lang/cargo/issues/8379 being fixed. :)

joking aside, there's so many code in our monorepo, which shouldn't be used in production code, but made `pub` only for the sake of usage in dependent crate's test/bench purposes.

when reviewing prs which touch such a method, it's too laborious to verify there's no misuse.

... and our social contract of naming things as `for_test(s)` and _be careful in general_ doesn't work always: 72919fcacc42713eccc4aaf91cb19bcd20dc8764 in the (reverted) commit, zk-token-proof is using `set_data`, which is indicated not to be used in production.

also, `#[deprecated]` won't work: the code should still be usable for test code; and sparkling `#[allow(deprecated)]`s at each and every call site would be too laborious.

#### Summary of Changes

Use a well-known work around: https://github.com/rust-lang/cargo/issues/8379#issuecomment-1261970561

The caveat is that some special treatment must be taken due to cargo's feature unification. so I added `cargo hack` tricks in `test-checks.sh`.

**the downside is that ci takes longer.... (don't scared of this pr's build time; this is mainly due to unpopulated `target` dir...; usually should take +~5 mins)**

see 72919fcacc42713eccc4aaf91cb19bcd20dc8764 (note that this is reverted) or https://github.com/solana-labs/solana/pull/32170#discussion_r1231833844 to see how's dev's life change after this feature debut.

lastly, this feature makes call-hierarchy-based auditing easier, binaries smaller, and last-not-least, humans lazier. lol

#### new pipeline example

image

**naming is finalized; please open follow up pr if you insist**

below is kept for historical reference

> #### Naming <div id="naming"/>
> - `dev`: too short?
> - `dev-utils`: uncommon phrase; inspired from `dev-dependencies` and they're closely related. not used in monorepo (i think introducing a new terminology is good)
> - `test-utils`: common phrase; sounds `bench` are excluded. already used in monorepo as mod name.. (i think this is bad because of some human's cognitive ambiguity.)
> - `non-essential`/`not-production`: a bit too abstract?
> - `dev-only`/`for-dev`: 2nd candidate?
> - `dev-profile-only-utils`: _dev-profile_ implies `-Copt-level=0` and etc for me (ref: https://doc.rust-lang.org/cargo/reference/profiles.html#dev). that's not true for benches and `cargo test --release` (`dev-utils` could be valid use cases for these)
> - `test-in-workspace`: very explicit name for the purpose of the feature. this is out of scope thinking but this new feature flag can theoretically be used by third-party downstream crates outside the monorepo. so better off avoiding `workspace`? btw, #31239 is using this.
> - `*-helper`?: imo, `helper` sounds some wrapper and less generic `utils`. we're planning to feature gate any dev related code under the new feature.
> - `sol{,ana}`?: needs proper justification.
> - `use-*`/`export-*`: i think these are redundant.
> - `{dev/test}-code`?: this as well redundant.
> 
> ##### utils vs util?
> 
> seems `utils` is popular?

### CODE REVIEW

**Comment by t-nelson** [comment]
looks like `cargo-hack` is currently trailing latest cargo by ten releases. any concerns with adding a tooling dependency they may break on us?

**Comment by ryoqun** [comment]
> looks like `cargo-hack` is currently trailing latest cargo by ten releases.

hmm, how do you think so? maybe, you saw MSRV at https://crates.io/crates/cargo-hack (currently 1.60)? it's just build dependency of `cargo-hack` itself for _minimum_ rust version.

it's released periodically: https://github.com/taiki-e/cargo-hack/releases

> any concerns with adding a tooling dependency they may break on us?

i think there's little concern.  `cargo-hack` is a thin wrapper for `cargo`. it's just reading rather stable Cargo.toml and  spawning actual `cargo` subcommands for ease.

**Comment by t-nelson** [comment]
> maybe, you saw MSRV

ah yeah, didn't realize that msrv was so prominent on crates.io nowadays. i probably just saw the first version-looking string and went with it :upside_down_face:

**Comment by t-nelson** [comment]
lol... run `cargo hack check --help` in the root of the monorepo :upside_down_face:

**Review by t-nelson — COMMENTED** [review]
i'm generally supportive of this change for the sake of eliminating test-only logic from crate public apis.  in order to fully take advantage, we'll need to replace all instances of `cargo {test,check}` with the `cargo-hack`-wrapped versions, right?

File: ci/test-checks.sh
Comment (ryoqun) [reply ci/test-checks.sh:32]: yeah, https://github.com/solana-labs/solana/pull/32234

File: ci/buildkite-pipeline.sh
Comment (yihau) [reply ci/buildkite-pipeline.sh:143]: yeah. I think we need to split it. for now check become a 17mins monster 🦖

**Comment by ryoqun** [comment]
> we'll need to replace all instances of `cargo {test,check}` with the `cargo-hack`-wrapped versions, right?

@t-nelson nope. as this pr introduced another set of `cargo hack check` invocations, there should be no additional benefit of adjusting existing cargo invocations. And `cargo hack test` will take too long...

File: ci/buildkite-pipeline.sh
Comment (ryoqun) [reply ci/buildkite-pipeline.sh]: this is done.

File: ci/test-checks.sh
Comment (ryoqun) [reply ci/test-checks.sh]: done

File: ci/test-checks.sh
Comment (ryoqun) [reply ci/test-checks.sh]: done

File: ci/test-checks.sh
Comment (ryoqun) [reply ci/test-checks.sh]: this is done.

**Comment by ryoqun** [comment]
oookay. I've finally finished up this pr. ready for review. also, I'll create bunch of test prs to test this pr.

File: scripts/check-dev-utils.sh
Comment (ryoqun) [inline scripts/check-dev-utils.sh:1]: unexpected from me? lol I've done some dense commenting work. :)

File: scripts/check-dev-utils.sh
Comment (ryoqun) [inline scripts/check-dev-utils.sh:31]: currently empty; i'll add some with a follow up pr.

File: scripts/check-dev-utils.sh
Comment (ryoqun) [inline scripts/check-dev-utils.sh:36]: hehe, I added this set of new checks after i drafted this pr.

File: scripts/check-dev-utils.sh
Comment (ryoqun) [inline scripts/check-dev-utils.sh:34]: the default `full` is intended for local development use.

**Comment by ryoqun** [comment]
> also, I'll create bunch of test prs to test this pr.
phew, ci abusing time \(^o^)/

I've prepared 6 scenarios and all results are expected.

### `1a`: https://github.com/solana-labs/solana/pull/32170

simplest case: some code is newly `dev-utils`-gated; but some affected crates aren't updated to activate `dev-utils`.

note that only `./check-dev-utils.sh check-all-targets` is failing as expected, justifying its existence.
that means existing `cargo check` is lying due to feature unification.

### `1b`: https://github.com/solana-labs/solana/pull/32171

`1a` is fixed with 6b33b13ba0868a7624926b69a3f78510ca5a8329 with additional dev-utils activation for the rest of needed crates.

### `2a`: https://github.com/solana-labs/solana/pull/32246

enabling `dev-utils` as part of normal dependencies is forbidden

note that only `./check-dev-utils.sh tree`'s first cargo tree is failing as expected, justifying its existence.

### `2b`: https://github.com/solana-labs/solana/pull/32247

`2a `is fixed with b04744794b3147e7aceb76e0837f61707f422779 with correct white listing

### `2c`: https://github.com/solana-labs/solana/pull/32248

`2b` was made to fail yet again with new bad dependent crate on the whitelisted crate.
to make it pass, it's needed to _contaminate_ any subtrees for `dev-utils`-as-normal-deps to damage control accidental `dev-utils` activation via feature unification.

note that only `./check-dev-utils.sh tree`'s second cargo tree is failing as expected, justifying its existence.

### `2d`: https://github.com/solana-labs/solana/pull/32249

`2c` is fixed with 4d561f18dcef9e1835a14d80b0877996ebecf4f2 with correct tainting.

### conclution

all in all, i think that almost all of `dev-utils` careless misuses will be detected while its usability and utility are upheld

**Comment by yihau** [comment]
does this PR mean that we need to add `dev-utils` feature to all test only functions?
if yes, what do you think about maybe we just move the impl to cfg(test). it looks like
<details>
<summary>code</summary>

```rust
struct S {
    v: u64,
}

impl S {
    fn foo(&self) -> u64 {
        self.v * 2
    }
}

#[cfg(test)]
impl S {
    fn bar(&self) -> u64 {
        self.v * 0
    }
}

fn main() {
    let s = S { v: 10 };
    println!("{:?}", s.foo());  // 20
    // println!("{:?}", s.bar()); // NOT ALLOW
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_works() {
        let s = S { v: 10 };
        assert_eq!(s.bar(), 0);
    }
}
```

</details>

**Comment by ryoqun** [comment]
> does this PR mean that we need to add `dev-utils` feature to all test only functions to avoid misusing?

yes. :)

> if yes, what do you think about maybe we just move the impl to cfg(test). it looks like

hehe, i understand this is natural question. but `cfg(test)` doesn't work across crates. your example is perfectly working because it's inside a single crate.

for details, see https://github.com/rust-lang/cargo/issues/8379. also, here's another use case for `dev-utils` in this monorepo: https://github.com/solana-labs/solana/pull/32190#discussion_r1237324036

File: scripts/check-dev-utils.sh
Comment (ryoqun) [inline scripts/check-dev-utils.sh:71]: here

**Comment by yihau** [comment]
you're right! I thought we only want to solve the question about misusing test only functions. if we want to also build a cross crates system in this PR. a feature hack seems the most popular way.

a no rush suggestion is that could we make the error message of check 1 more easy to read? maybe just tell them how to do instead of the some inverted trees haha.

would like to wait for Trent's green light 😈

**Review by t-nelson — COMMENTED** [review]
you've nerd-sniped me into attempting to port this `cargo tree` (brittle output) + `grep` (fuzzy) logic into `cargo metadata` (stable) + `jq` (concise). if i succeed, i think we can offer more clear guidance messages upon failure

File: scripts/check-dev-utils.sh
Comment (t-nelson) [reply scripts/check-dev-utils.sh:10]: actually maybe a `use-` or `export-` would better self-describe as well :thinking:

Comment (ryoqun) [reply scripts/check-dev-utils.sh:10]: > can we call the feature `test-utils` instead? or something that implies that the feature gates test-specific logic.

well i avoided `test` because of existence of `benches`. so, i followed cargo's `dev-dependency` (it isn't called `test-dependency`).

>  maybe `sol{,ana}`-prefixed

hmm, i don't feel strong for prefixing as other existing features aren't prefixed. also, features are usually namespaced to workspace. so, even if we use `--features XXX`, it won't enable features from other crates. say, `--features serde` on CLI doen't enable `im`'s `serde` feature. so, I wonder what's justification of prefixing just for `dev-utils` in that regard too.

> actually maybe a `use-` or `export-` would better self-describe as well :thinking:

i think features themselves imply _using_ / _exporting_ as cargo's features are additive by design. this new feature (`dev-utils`) works in additive manner just like all other normal features. we're just special-casing it to prevent accidental overuse and misuse at CI layer.

File: scripts/check-dev-utils.sh
Comment (ryoqun) [reply scripts/check-dev-utils.sh:34]: done: 2b9940ea919020663d2e9cd8027ce43fcd70e265

**Comment by t-nelson** [comment]
> you've nerd-sniped me into attempting to port this `cargo tree` (brittle output) + `grep` (fuzzy) logic into `cargo metadata` (stable) + `jq` (concise). if i succeed, i think we can offer more clear guidance messages upon failure

fwiw, i think this is going to work. i got the first check and most of the second reimplemented friday evening. i'll try to wrap it up this coming week and get a branch out for you to review

File: scripts/check-dev-utils.sh
Comment (t-nelson) [reply scripts/check-dev-utils.sh:10]: oh _that_ `dev-` :sweat: 

i think we need another word in the middle to disambiguate. will think on it some more

File: scripts/check-dev-utils.sh
Comment (t-nelson) [reply scripts/check-dev-utils.sh:34]: `case` would be cleaner, but this works!

**Comment by t-nelson** [comment]
ok. here's what i came up with. lmk if it looks like it's on the right track and i'll cleanup variable names and terminology

https://github.com/ryoqun/solana/compare/dev-utils...t-nelson:solana:dev-utils-tree+jq

**Comment by ryoqun** [comment]
@t-nelson cool, it looks good at quick glance. seems quite good example to learn `jq` power. :) so, what to do shipping this `dev-utils`? push the jq commit to here? or, merge this and create follow up by you with the jq commit? i don't bother, just want to ship this thing quickly. :)

fyi, I'm going to create bunch of small prs which actually starts to use `dev-utils` for others to learn how to use it.

> i think we need another word in the middle to disambiguate. will think on it some more

oh, maybe, this pr is blocked on this? did you came up better name by any chance?

**Comment by t-nelson** [comment]
> @t-nelson cool, it looks good at quick glance. seems quite good example to learn `jq` power. :)

:mortar_board: 

> so, what to do shipping this `dev-utils`? push the jq commit to here? or, merge this and create follow up by you with the jq commit? i don't bother, just want to ship this thing quickly. :)

if you wanna pull them in here and just squash them, that's fine by me. i don't need credit :wink:

> fyi, I'm going to create bunch of small prs which actually starts to use `dev-utils` for others to learn how to use it.

:tada: 

> > i think we need another word in the middle to disambiguate. will think on it some more
> 
> oh, maybe, this pr is blocked on this? did you came up better name by any chance?

wdyt of `dev-profile-only-utils`?  a bit long, but it's a short as i could make concise

also wondering if we should name the ci marker feature more generically? i could see us wanting to do other custom things under ci. `solana-ci-in-effect` or maybe just `solana-ci`?

**Comment by ryoqun** [comment]
> > so, what to do shipping this `dev-utils`? push the jq commit to here? or, merge this and create follow up by you with the jq commit? i don't bother, just want to ship this thing quickly. :)
> 
> if you wanna pull them in here and just squash them, that's fine by me. i don't need credit wink

okay, i'll pull them in here and will finish up the jq approach.



> > > i think we need another word in the middle to disambiguate. will think on it some more
> > 
> > 
> > oh, maybe, this pr is blocked on this? did you came up better name by any chance?
> 
> wdyt of `dev-profile-only-utils`? a bit long, but it's a short as i could make concise

hmm, _dev-profile_ implies `-Copt-level=0` and etc for me (ref: https://doc.rust-lang.org/cargo/reference/profiles.html#dev). that's not true for benches and `cargo test --release` (`dev-utils` could be valid use cases for these)

> also wondering if we should name the ci marker feature more generically? i could see us wanting to do other custom things under ci. `solana-ci-in-effect` or maybe just `solana-ci`?

i think we're better off not making the ci marker generic at least for now. i think overloading a common marker feature for various things could lead to false positives/negatives. fyi, i have sneaked in another one called `dummy-for-ci-check` at #30602. but i can't unify the two due to different requirments. 

crate features are rather less pronounced when using it (docs.rs, crates.io). so, bunch of mysterious features won't hurt much, i guess.

**Comment by ryoqun** [comment]
> > if you wanna pull them in here and just squash them, that's fine by me. i don't need credit wink
> 
> okay, i'll pull them in here and will finish up the jq approach.

I've done the polishing. and pushed all commits. thus, requesting review @t-nelson 

> > also, I'll create bunch of test prs to test this pr.
> 
> phew, ci abusing time (^o^)/
> 
> I've prepared 6 scenarios and all results are expected.

I've re-run those (thanks @yihau for quick #32340 addressing)

seems all are working as expected.

**Comment by ryoqun** [comment]
well, i spotted a bug and fixed [652df19](https://github.com/solana-labs/solana/pull/32169/commits/652df19cbcc71a74b44819efc07e1a55a7d525d3)

**Review by t-nelson — COMMENTED** [review]
> hmm, dev-profile implies -Copt-level=0 and etc for me (ref: https://doc.rust-lang.org/cargo/reference/profiles.html#dev). that's not true for benches and cargo test --release (dev-utils could be valid use cases for these)

haha... dammit. i want a general word for this context

Screenshot from 2023-06-30 19-00-54

**Comment by t-nelson** [comment]
> haha... dammit. i want a general word for this context

oh durr... `dev-context-only-utils` ?

File: scripts/check-dev-utils.sh
Comment (ryoqun) [reply scripts/check-dev-utils.sh:96]: https://github.com/solana-labs/solana/pull/32169#issuecomment-1614708753 has little bit of context. 

and i added `2e` and `2f` scenarios here: https://github.com/solana-labs/solana/pull/32169#issuecomment-1603932979

File: scripts/check-dev-utils.sh
Comment (ryoqun) [reply scripts/check-dev-utils.sh]: yeah, i know. i wanted to be intended like this rust counterpart...: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9b9fbdc76d8d3d2dcf3d92bb1a689bae

wish bash's `<<-` recognize spaces (i hate tabs)....

reverted from odd `echo`-ing..: 58013c2a12c3fe9f73da64a8496ff262ec95fd00

File: scripts/check-dev-utils.sh
Comment (ryoqun) [reply scripts/check-dev-utils.sh]: done: https://github.com/solana-labs/solana/pull/32169#discussion_r1248493752

**Comment by ryoqun** [comment]
> > haha... dammit. i want a general word for this context
> 
> oh durr... `dev-context-only-utils` ?

allow me to defend `dev-utils` a bit...

hmm, I wonder the `-context-only-` could convey the subtle implication of what you wanted to express. and the name is longer (considerably, imo). both `context` and `only` are generic in their meaning. so, i  couldn't derive some valuable information from those words. say, `serde-only` vs just `serde` and `alloc-context-only-utils` and `alloc-utils`.

also, as we're recognizing usage of `dev-utils` as normal dep for benches, i'm not sure about the merit of indicating that context so vocally in its name... for that matter, `dev-utils` is defined and used just like normal features like `serde`s. and all of features including `dev-utils` are govenred in the same rule like no-propagation under dev dep. usage...

lastly, we have (hopefully) proper ci guard against misuse.

File: scripts/check-dev-utils.sh
Comment (ryoqun) [reply scripts/check-dev-utils.sh]: nice :) https://github.com/solana-labs/solana/pull/32169/commits/bc215ae73a08d5f78c5a9fb95344049f0fde6f1c

File: scripts/check-dev-utils.sh
Comment (t-nelson) [reply scripts/check-dev-utils.sh]: hehe. yeah necessary evil imo. i do cringe at heredoc indent not following the current scope :grimacing:

**Comment by t-nelson** [comment]
the main reason i don't believe `dev-utils` to be sufficient is i suspect that most people will follow logic along the lines of "`dev-utils`. i'm a dev. this is a utility. this needs `dev-utils`", then get hammered by CI and now we have to support them on discord. given that rust folks have been zero help in giving us a nice concise name, the only sane alternative is to trigger a "wtf is this?" to (hopefully!) prompt a docs search upon first exposure

**Comment by ryoqun** [comment]
> the main reason i don't believe `dev-utils` to be sufficient is i suspect that most people will follow logic along the lines of "`dev-utils`. i'm a dev. this is a utility. this needs `dev-utils`", then get hammered by CI and now we have to support them on discord. given that rust folks have been zero help in giving us a nice concise name, the only sane alternative is to trigger a "wtf is this?" to (hopefully!) prompt a docs search upon first exposure

@t-nelson thanks for clarification. I'm persuaded and giving up shorter name: 6e6d1897eab0255d78dd1dd3e66a45e27b1966d4

requesting hopefully final review. :) Also, I'll update the ci scenarios prs shortly after :) (hint: i wonder it can make it to ship this today before US holiday...)

**Review by t-nelson — APPROVED** [review]
woot!  let's give it a shot

:shipit:

**Pull request merged**
